plugin-server/src/cdp/exceptions-manager.ts

```diff
@@ -0,0 +1,101 @@
+import LRUCache from 'lru-cache'
+
+import { Hub, Team } from '../types'
+import { PostgresUse } from '../utils/db/postgres'
+import { GroupType, HogFunctionInvocationGlobals } from './types'
+
+export type GroupsMap = Record<string, GroupType>
+export type GroupsCache = Record<Team['id'], GroupsMap>
+
+// Maps a fingerprintfor easy lookup like: { 'team_id:merged_fingerprint': primary_fingerprint }
+type ExceptionFingerprintByTeamType = Record<string, string>
+
+const FINGERPRINT_CACHE_AGE_MS = 60 * 10 * 1000 // 10 minutes
+
+export class GroupsManager {
+    fingerprintMappingCache: LRUCache<number, Record<string, string[]>> // team_id: { primary_fingerprint: merged_fingerprints[] }
+
+    constructor(private hub: Hub) {
+        // There is only 5 per team so we can have a very high cache and a very long cooldown
+        this.fingerprintMappingCache = new LRUCache({ max: 1_000_000, maxAge: FINGERPRINT_CACHE_AGE_MS })
+    }
+
+    private async fetchExceptionFingerprintMapping(teams: Team['id'][]): Promise<ExceptionFingerprintByTeamType> {
+        const exceptionFingerprintMapping: ExceptionFingerprintByTeamType = {}
+
+        // Load the cached values so we definitely have them
+        teams.forEach((teamId) => {
+            const cached = this.fingerprintMappingCache.get(teamId)
+
+            if (cached) {
+                Object.entries(cached).forEach(([primaryFingerprint, mergedFingerprints]) => {
+                    mergedFingerprints.forEach((mergedFingerprint) => {
+                        exceptionFingerprintMapping[`${teamId}:${mergedFingerprint}`] = primaryFingerprint
+                    })
+                })
+            }
+        })
+
+        const teamsToLoad = teams.filter((teamId) => !this.fingerprintMappingCache.get(teamId))
+
+        if (teamsToLoad.length) {
+            const result = await this.hub.postgres.query(
+                PostgresUse.COMMON_READ,
+                `SELECT fingerprint, merged_fingerprints, team_id
+                FROM posthog_errortrackinggroup
+                WHERE team_id = ANY($1) AND merged_fingerprints != '{}'`,
+                [teamsToLoad],
+                'fetchExceptionTrackingGroups'
+            )
```

Comment:
This is the part that I simplified....

We could lookup based on `$exception_fingerprint` as another clause in this where (something akin to `merged_fingerprints @> ANY('{$1}'::text[])` where $1 represents an array of the fingerprints). We'd then have to normalise the rows to a `team_id:merged_fingerprint => primary_fingerprint` combo and only cache the ones originally passed into the function. We would have to add multiple clauses to make sure that the fingerprints were scoped to the relevant team.

I opted to keep things simple at the expense of a bigger cache. We lookup the teams instead and find all groups with a "merged fingerprints". We add the mapping for all merged fingerprints to the cache, irregardless of whether the group is needed for the current set of `items`

Reply:
This could be a problem down the line if the cache starts to grow because teams have a large number of exception groups with merged fingerprints. Because we need to merge _all_ fingerprints regardless of whether they are archived or not the response for a team will only grow over time.

In practice I'm not too worried because merging is a deliberate action in the UI so the numbers should be small initially. In future we might have to trade the query complexity off to reduce the number of returned rows (thus having a smaller cache that only contains more recent fingerprints seen for a team)